### DISCUSSION TITLE
ESP32 Firmware bin file for micro python?

### DISCUSSION THREAD

**Discussion by EtherealNate** [micropython/micropython#14353] [ESP32]
Hello everyone.  I have been unable to locate the firmware for my esp32 board.  I have spent a lot of time searching and I have hit a brick wall.  I would really appreciate being pointed in the right direction.  The board is labelled wroom-32 cp2102 it is dual core with a usb c interface.  What is the correct firmware?

**Comment by shariltumin** [comment]
You can try https://micropython.org/download/ESP32_GENERIC/

Click on the v1.22.2 (2024-02-22) .bin  to download.

**Reply by EtherealNate** [reply]
I downloaded that file but I am having trouble flashing the firmware.  It seems that this chip doesn't support usb c for flashing, only UART.  Is that correct?

**Reply by scruss** [reply]
What are you using to flash it? 
Does your system need a driver for the CP2012 serial chip?

**Reply by EtherealNate** [reply]
I might need the driver.  I am using a usb c plug and the flash download tool 3.9.6.  I am thinking I need to config the plug for uart to flash the firmware.  When I try to flash the firmware I get null in the cmd.  Thoughts?

**Reply by shariltumin** [reply]
The cp2120 is usb to uart, so I think you need to flash via UART using **esptool.py**. As Stewart said, you may need the cp2021 driver if you are running Windows.

**Reply by EtherealNate** [reply]
That sounds like a great next step.  Thanks.

**Reply by EtherealNate** [reply]
It is actually 2102.  I typed the numbers in the wrong order.

**Reply by EtherealNate** [reply]
I believe I have all the drivers that I need.  I am just waiting on the right adapter.  I should have it in 2 days.  Thanks again for all the help!

**Reply by EtherealNate** [reply]
I have changed my approach.  I am now using a raspberry pi with my UART adapter to flash the firmware onto my esp32.  I am getting a write_flash error: Address must be a number.  Is it an order issue or do I have the wrong memory address?  I was looking through the documentation for the correct address but to no avail.  What am I missing?

**Reply by shariltumin** [reply]
I assume you are using esptool.py to flash the firmware. I also assume the firmware file has been renamed to firmware.bin, then please try:

```bash
esptool.py --port /dev/ttyUSB0 --chip esp32 --baud 460800 write_flash -z 0x1000 firmware.bin

```

Once you have connected the USB-UART adapter to your Raspberry Pi, do a ```ls /dev/ttyUSB*```. Or you can use ```dmesg | grep tty```. This is the "--port" device you should be using.

**Reply by EtherealNate** [reply]
I keep getting the error: Address must be a number.

**Reply by shariltumin** [reply]
Can you please send me the command you are using and the output?

**Reply by EtherealNate** [reply]
I have tried a few variations.  esptool.py --port /dev/ttyUSB0 write_flash -z 0x1000 /path/to /firmware.bin.  I used "esptool.py --port /dev/ttyUSB0 --chip esp32 --baud 460800 write_flash -z 0x1000 firmware.bin".  I am also having trouble erasing the firmware.  I am getting no serial data received.  Is it because the chip isn't in boot mode or is the chip defective?  I have the 5v pin connected, should I use 3.3v?

**Reply by shariltumin** [reply]
Can you try:

```
esptool.py --port /dev/ttyUSB0 --chip esp32 --baud 460800 --trace flash_id
```

Are you seeing error messages?

Do you have read and write permissions for /dev/ttyUSB*? Which groups does your account belong to? You can try running the command with sudo.

**Reply by EtherealNate** [reply]
Now i am getting  failed to connect to esp32: invalid head of packet(0x00): Possible serial noise or corruption.  I ran the trace flash_id and I get read invalid data and remaining data in serial buffer.

**Reply by EtherealNate** [reply]
I decided to try a new tactic.  I was able to connect to my esp32 with a usb-c adapter.  I think my uart is defective.  I will explore that another time.  I was able to flash new firmware and now it is almost ready for code.  I think my firmware that I am using is incomplete.  I am using jumper wires for I2C with a small oled screen.  I want to display a simple message but it doesn't recognize the screen.  I ran the cmd i2cdetect -y 1 and the address for the screen doesn't show up.  I am looking for a more complete firmware that would include the network, machine and any other I2C modules.

**Reply by EtherealNate** [reply]
I was able to get my board working.  I uploaded a simple code and so far everything works!  Thank you for all the help.  I still have some puzzling data to go through but for now I will take this victory.  I will leave this discussion open for a while longer incase anyone has something else to add.  Until my next question, happy programming!

**Reply by shariltumin** [reply]
I take my victories, however small. I welcome my failures as lessons learned. Nothing is so complicated that we cannot break it down into small, manageable pieces. I'm glad you haven't given up.
